### ISSUE TITLE
Opening detail page for quarantined messages takes very long - releasing them also

### ISSUE THREAD

**Issue by Stephanowicz** [mailwatch/1.2.0#1039]
Hi,

we use MailWatch für MailScanner v1.2.7 on Debian stretch x64 running on a really fast hyper-v machine.

But when opening a detail page for quarantined messages it takes aprox > 20 sec til the page opens.
Releasing such a mail takes even longer.

Opening a detail page for a normal mail is fast.

I found no errors in the log files...
Anything else I can check? Or is this the expected behaviour?

Cheers,

Stephan

**Comment by Skywalker-11** [comment]
It shouldn't be much slower.
One reason for it to be slower could be slow storage. 
Mails that aren't quarantined only access the database and not the disk while qurantined mails require both database and file accesses.
An other reason could be that the mail was quarantined on a different server/servers localhost doesn't match the quarantining server. In that case additionally to the first point it requires an rpc connection to the other server and file access there.

Please also provide your system info as mentioned here: https://github.com/mailwatch/MailWatch/blob/1.2/.github/ISSUE_TEMPLATE.md

**Comment by Stephanowicz** [comment]
Hi,

ok - here some specs:

Mailwatch: https://github.com/mailwatch/MailWatch/archive/v1.2.7.tar.gz

Debian stretch 4.9.0-5-amd64 #1 SMP Debian 4.9.65-3+deb9u2 (2018-01-04) x86_64 GNU/Linux
mailscanner 5.0.7-1
mariadb-server 10.1.26-0+deb9u1
postfix 3.1.6-0+deb9u1 amd64
PHP 7.0.27-0+deb9u1 (cli)
Apache2 2.4.25-3+deb9u3 amd64

System: 
Kernel Version 4.9.0-5-amd64 (SMP) x86_64 
4x Intel(R) Xeon(R) Silver 4108 CPU @ 1.80GHz

Client: Win10 IE11 || FF58.0.2

**Comment by Stephanowicz** [comment]
-Mailwatch/Mailscanner etc are on the same host
-storage speed is more then enough

is there any way to debug this, or to get some further info why this takes so long?

Thanks, Stephan

**Comment by Skywalker-11** [comment]
@Stephanowicz Does it take so long for every mail that is quarantined or only some of them? Eg. if you send yourself a mail with an attached virus? 
It could take longer if MW tries reverse lookup the sending server. If that fails it may take longer to load the page. But the same should happen with unquarantined mails that can't be reverse lookuped.

**Comment by Stephanowicz** [comment]
Opening a detail page for **any** quarantained mail takes aprox 20 seconds - the same for releasing them 

You mean, I should check if it is a resolving issue, by sending a infected mail from a local account?
I can test this.

Any options for debugging?

In what way mailscanners is involved in this? Could I do some debugging there?

**Comment by Stephanowicz** [comment]
O.K., just checked it with a local mail account - the details page is opening in less then a second... :)

So, what can I do about this?

**Comment by Stephanowicz** [comment]
@Skywalker-11 

> It could take longer if MW tries reverse lookup the sending server. If that fails it may take longer to load the page. But the same should happen with unquarantined mails that can't be reverse lookuped.

Yes, this also happens to unquarantined mails - I get  `(Reverse Lookup fehlgeschlagen) `

These are mostly private IPs

**Comment by Skywalker-11** [comment]
@Stephanowicz try to comment https://github.com/mailwatch/MailWatch/blob/87ccad3452b9e0b9b1ac2ec8d248f980d4838ec5/mailscanner/detail.php#L177-178
Change
```
                    } elseif ($isLocalNetwork === true) {
                        $output .= ' <td>' . __('localhost04') . "</td>\n";
                    } elseif ($geoip_country = return_geoip_country($relay)) {
                        $output .= ' <td>' . $geoip_country . '</td>' . "\n";
                    } else {
                        $output .= ' <td>' . __('geoipfailed04') . '</td>' . "\n";
                    }
```
to
```
                       } elseif ($isLocalNetwork === true) {
                        $output .= ' <td>' . __('localhost04') . "</td>\n";
/*                    } elseif ($geoip_country = return_geoip_country($relay)) {
                        $output .= ' <td>' . $geoip_country . '</td>' . "\n"; 
*/
                    } else {
                        $output .= ' <td>' . __('geoipfailed04') . '</td>' . "\n";
                    }
```

You won't get any country lookup anymore but it should be faster. 

As we also have to replace the geoip lookup soon (#1037) we should take a look at the timing aspect while doing that

**Comment by Stephanowicz** [comment]
Nice,

doing so reduced the time to aprox 6-8 sec
This lead me to look for `gethostbyaddr()` and I found  this neat function:
```

//faster alternative to gethostbyaddr()
private function gethost( $ip )
{
    //Make sure the input is not going to do anything unexpected
    //IPs must be in the form x.x.x.x with each x as a number

    if( preg_match( '/^(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)(?:[.](?:25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)){3}$/', $ip ) )
    {
        $host = `host -s -W 1 $ip`;
        $host = ( $host ? end( explode( ' ', trim( trim( $host ), '.' ) ) ) : $ip );
        if( in_array( $host, Array( 'reached', 'record', '2(SERVFAIL)', '3(NXDOMAIN)' ) ) )
        {
            return sprintf( '(error fetching domain name for %s)', $ip );
        }
        else
        {
            return $host;
        }
    }
    else
    {
        return '(invalid IP address)';
    }
}
```
([https://stackoverflow.com/a/8729116](https://stackoverflow.com/a/8729116))

I modified it a bit to act more like the original `gethostbyaddr() `and put it to `funtions.php ` - now it takes only about 2 secs :)
```

//faster alternative to gethostbyaddr()
function gethost( $ip )
{
    //Make sure the input is not going to do anything unexpected
    //IPs must be in the form x.x.x.x with each x as a number

    if( preg_match( '/^(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)(?:[.](?:25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)){3}$/', $ip ) )
    {
        $host = `host -s -W 1 $ip`;
        $host = ( $host ? end( explode( ' ', trim( trim( $host ), '.' ) ) ) : $ip );
        if( in_array( $host, Array( 'reached', 'record', '2(SERVFAIL)', '3(NXDOMAIN)' ) ) )
        {
           // return sprintf( '(error fetching domain name for %s)', $ip );
              return $ip;
        }
        else
        {
            return $host;
        }
    }
    else
    {
        //return '(invalid IP address)';
          return false;
    }
}
```

**Comment by Stephanowicz** [comment]
Maybe one could add a button or similar on the details page to lookup geoip on demand - so that it isn't done by default - only if needed....

Cheers, Stephan

**Comment by Skywalker-11** [comment]
@Stephanowicz for the 1.2 branch we won't add additional features but only bug fixes. But I will note this for 2.x release which is currently beeing developed

**Comment by thctlo** [comment]
Hai, small breakin here.. 

@Stephanowicz  
You did check your apache apache2/error.log and check your mysql error logs. 

I noticed simular thing but that was due to an email attachment the expanded over 128Mb ( my php max mem limit ) 

Can you provide the exact steps to test this because i dont see these delays. 
Openen a details page take about 1 sec for me and same for the release. 
Running Debian stretch the same as you with a very restrictive firewall and in addition bind9.

When its all done i'll share the howto so everybody can enjoy debian with mailscanner+mailwatch. ;-)
im at 99%, then a reply of the howto to test it. 

Guest host running on Xen Server 7.2. 
model name   : AMD Opteron(tm) Processor 4386
cpu MHz         : 3099.865
VM set to 4 cores an 4GB ram.

Try this also, set swappiness to 10-20 
This make the system use up to 80-90% of all ram before swapping

`/etc/sysctl.d/10-swappiness.conf
cat /proc/sys/vm/swappiness
#Default: 60
vm.swappiness=20
`

run : sysctl -p /etc/sysctl.d/10-swappiness.conf

Sidenote here. 
These slownessess are often a fault in resolving/dns/routing setups, or firewalling things. 

I also use a caching + forwarding bind setup on my mailscanner also. 
a zone with forward to my lan dns server ( internal interface ) 
a zone with forward to my internet dns server ( external interface ) 
This was needed also for my SPF DKIM TSLA CAA/DANE integration in postfix.

**Comment by Skywalker-11** [comment]
@thctlo The delay mentioned here most likely occures if the reverse ip address of one/multiple servers mentioned in the mail headers can't be resolved by geoip and/or the dns servers. At least when I was investigating this issue it only affected mails where reverse lookup fails. Eg. I can reproduce the delay with the following mail:
```
Received-SPF: Softfail (mailfrom) identity=mailfrom; client-ip=115.59.74.242; helo=gmail.com; envelope-from=patrhoyt29jteci@gmail.com; receiver=<UNKNOWN>
Received: from gmail.com (unknown [115.59.74.242])
        by xxxxx (Postfix) with SMTP id A39FF9FD5E
        for <xxxxx>; Sun,  7 Jan 2018 12:27:43 +0100 (CET)
Received: from unknown (170.11.42.44)
        by smtp-server1.cfdenselr.com with LOCAL; Sun, 07 Jan 2018 00:10:58 -1100
Received: from m1.gns.snv.thisdomainl.com ([Sun, 07 Jan 2018 00:03:41 -1100])
        by public.micromail.com.au with NNFMP; Sun, 07 Jan 2018 00:03:41 -1100
Received: from unknown (HELO smtp18.yenddx.com) (Sat, 06 Jan 2018 23:49:42 -1100)
        by mxs.perenter.com with ESMTP; Sat, 06 Jan 2018 23:49:42 -1100
Received: from qnx.mdrost.com ([209.171.75.40]) by smtp.endend.nl with LOCAL; Sat, 06 Jan 2018 23:42:54 -1100
Received: from [195.224.81.152] by mail.gimmicc.net with SMTP; Sat, 06 Jan 2018 23:30:51 -1100
Message-ID: <A0DDB600.50A83D3C@gmail.com>
Date: Sat, 06 Jan 2018 23:11:49 -1100
Reply-To: "Negative SEO works" <patrhoyt29jteci@gmail.com>
From: "Negative SEO works" <patrhoyt29jteci@gmail.com>
User-Agent: Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.8.1.3) Gecko/20070326 Thunderbird/2.0.0.0
MIME-Version: 1.0
To: <xxxx>
Subject: Protect your xxxx ranks - Fight back
Content-Type: multipart/mixed;
        boundary="------------547623216533105835686543"

This is a multi-part message in MIME format.

--------------547623216533105835686543
Content-Type: text/plain;
        charset="iso-8859-1"
Content-Transfer-Encoding: 7bit


Here is how
http://www.black-seo.org/

Regards
Negative SEO works














http://www.blackhatseos.com/unsubscribe/


--------------547623216533105835686543
Content-Type: application/octet-stream;
        name="SEOReport.html"
Content-Transfer-Encoding: base64
Content-Disposition: attachment;
        filename="SEOReport.html"

PGh0bWw+DQoNCjxoZWFkPg0KPE1FVEEgSFRUUC1FUVVJVj0iUmVmcmVzaCIgY29udGVudD0gIjI7
VVJMPWh0dHA6Ly93d3cuYmxhY2stc2VvLm9yZy8iPg0KPC9oZWFkPg0KDQo8cD48Zm9udCBmYWNl
PSJWZXJkYW5hIj5Mb2FkaW5nIDxiPjxhIGhyZWY9Imh0dHA6Ly93d3cuYmxhY2stc2VvLm9yZy8i
PlNFTyBSZXBvcnQgTG9hZGluZzwvYT4gPC9iPjwvZm9udD48L3A+DQo8L2h0bWw+DQoNCg==

--------------547623216533105835686543--
```

**Comment by Stephanowicz** [comment]
@thctlo  it's like skywalker said: there are ip-adresses in the header that can't be resolved and the described solution is working fine

cheers, stephan

**Comment by thctlo** [comment]
Yes, but that confirms its an configuration error in the resolving setup. 
and that test is funky test. 

Received: from gmail.com (unknown [115.59.74.242])
gmail.com is NOT 115.59.74.242
same for : 
Received: from unknown (170.11.42.44)

RFC states, any smtp cummunication should resolve to A or PTR or MX. 
https://tools.ietf.org/html/rfc5321 see : 2.3.5.  Domain Names 
Only resolvable, fully-qualified domain names (FQDNs) are permitted
   when domain names are used in SMTP.  In other words, names that can
   be resolved to MX RRs or address (i.e., A or AAAA) RRs 

Configure your setup better, you should never allow this.
This is asking for problems in the long run, but thats my opinion. 

If everything is correctly configured you dont have the slowdown. 

Not a bug, but error in configuration, because i can not reproduce this  on my setup.

**Comment by Skywalker-11** [comment]
@thctlo thanks for pointing that out. 
Do you check all servers that are mentioned in the headers when receiving a mail or only the server you are directly communicating with? 
If not all servers mentioned in the headers are checked it could still delay showing the mail. Also the servers that the mail was relayed over may use custom/private dns servers that can resolve the ip/domain while they aren't resolvable through global dns. And so the lookup on your own server fails while the RFC statement still is valid for the other relaying server.

**Comment by Stephanowicz** [comment]
@thctlo well, in my case these adresses that can't be resolved are private enterprise adresses from internal mailservers & relays - other public adresses seem to resolve well as far as I know.

**Comment by thctlo** [comment]
@Skywalker-11 yes, i check every server in the headers.
> Also the servers that the mail was relayed over may use custom/private dns servers that can resolve the ip/domain while they aren't resolvable through global dns. And so the lookup on your own server fails while the RFC statement still is valid for the other relaying server.

The RFC is not speaking of private ranges, or i missed something here, ( can be, i dont know everything).. 

@Stephanowicz 
But that is cleary a misconfiguration, maybe you should ask yourself, why your mail host is not resolved for the internal side.  And since the solution is very simple...  I really suggest you try this out. 
and post us the result. 

Your setup is the same as mine, i'll give my bind9 setup as example, howto set this up. 
First, check if your firewall is setup to allow port 53 udp and tcp.  YES you need tcp also. ! 

To reduce the delay timeout for UDP connections, and thus highlight the randomization, which by default is 30s
`sysctl -w net.netfilter.nf_conntrack_udp_timeout=10`

bind9 
named.conf.options
```
acl localclients {
    localhost;
    localnets;
};
controls {
        inet 127.0.0.1 port 953 allow { 127.0.0.1; };
        inet ::1 port 953 allow { ::1; };
};
options {
        directory "/var/cache/bind";
        forwarders { 62.212.131.101; 8.8.8.8; };
        forward only;
        dnssec-enable yes;
        dnssec-validation yes;
        auth-nxdomain no;    # conform to RFC1035
        listen-on-v6 { ::1; };
        listen-on { 127.0.0.1; };
        min-cache-ttl 60;
        allow-query { localclients; };
        allow-query-cache { localclients; };
        recursion yes;
        allow-recursion { localclients; };
        allow-transfer { none; };
        max-cache-size 32M;
        empty-zones-enable no;
};

```

And named.conf.local
```
zone "internal.domain.tld" {
    type forward;
    forwarders { 192.168.1.1; 192.168.1.2; };
};

zone "domain.tld" {
    type forward;
    forwarders { 8.8.8.8; 62.212.131.101; };
};

```

And last, the resolv.conf
/etc/resolv.conf
```
search internal.domain.tld domain.tld
nameserver 127.0.0.1
# These are als set in /etc/bind/named.conf.options
nameserver 62.212.131.101
nameserver 8.8.8.8
# optional change one of the 2 internet ips to one of the internal dns servers.
```

The internal.domain.tld is the "primary" internal zone, pointed to my AD DC dns servers.  (samba4 AD DC dns servers.)
The domain.tld is the internet zone, pointed to some internet DNS servers. 

In above example, you only need to change 2 things.  
The domain names  and the ip adress of the dns servers. 

If you have questions, you can ask me,  no problem at all, we are here all to learn.

**Comment by Skywalker-11** [comment]
From the introduction of the RFC:
```
 An important feature of SMTP is its capability to transport mail
   across multiple networks, usually referred to as "SMTP mail relaying"
   (see Section 3.6).  A network consists of the mutually-TCP-accessible
   hosts on the public Internet, the mutually-TCP-accessible hosts on a
   firewall-isolated TCP/IP Intranet, or hosts in some other LAN or WAN
   environment utilizing a non-TCP transport-level protocol.  Using
   SMTP, a process can transfer mail to another process on the same
   network or to some other network via a relay or gateway process
   accessible to both networks.
```
Also 
```
2.3.4.  Host

   For the purposes of this specification, a host is a computer system
   attached to the Internet (or, in some cases, to a private TCP/IP
   network) and supporting the SMTP protocol.  Hosts are known by names
   (see the next section); they SHOULD NOT be identified by numerical
   addresses, i.e., by address literals as described in Section 4.1.2.

```

This means that it is completely valid to have private smtp servers that may are not resolvable from the internet but only in the internal network. These servers then could communicate with a relay server that has access to the internal dns (can resolve the internal FQDN) and faces the internet and is also resolvable on the internet. The internal server then can add a "Received: " line and the relay only adds his own "Received:" line (keeping the internal one) and both "Received:" lines will be send to your server.
The 2.3.5 section of the RFC matches this scenario as the relay server can validate the fqdn of the internal server.

Also notice that section 3.6.3 and  3.7.2 which explicitely states ` a relay SMTP has no need to inspect or act upon the header section or body of the message data and MUST NOT do so` and `the most important use of Received: lines is for debugging mail faults, and this debugging can be severely hampered by well-meaning gateways that try to "fix" a Received: line.... receiving systems MUST NOT reject mail based on the format ...`
If I understand this correctly you are violating the RFC if you reject mails based on the "invalid" header field.

Please correct me if I'm wrong

**Comment by thctlo** [comment]
`If I understand this correctly you are violating the RFC if you reject mails based on the "invalid" header field.`
Depends, if you do header checks and reject at MTA, yes.. but we dont do that.

**Comment by Stephanowicz** [comment]
@thctlo - thank You, but we have no problem resolving our intranet adresses - the problem arose from the privat **sender**-adresses in the mail header
And the solutions mentioned above are working very well!

**Comment by Skywalker-11** [comment]
MW 2.0 should request this information via ajax or similar so that the page load fast and only the "slow" information (here dns query) gets delayed
